drivers/python/rethinkdb/_export.py

```diff
@@ -188,12 +205,21 @@ def write_table_metadata(progress, conn, db, table, base_path):
 # This is called through rdb_call_wrapper and may be called multiple times if
 # connection errors occur.  In order to facilitate this, we do an order_by by the
 # primary key so that we only ever get a given row once.
-def read_table_into_queue(progress, conn, db, table, pkey, task_queue, progress_info, exit_event):
+def read_table_into_queue(progress, conn, db, table, pkey, keys, index, task_queue, progress_info, exit_event):
     read_rows = 0
-    if progress[0] is None:
-        cursor = r.db(db).table(table).order_by(index=pkey).run(conn, time_format="raw", binary_format='raw')
+    table = r.db(db).table(table)
+
+    if progress[0] is not None:
+        table = table.between(progress[0], None, left_bound="open")
+
+    if keys is None:
+        table = table.order_by(index=pkey)
+    elif index is not None:
+        table = table.get_all(*keys, index=index).order_by(pkey)
     else:
-        cursor = r.db(db).table(table).between(progress[0], None, left_bound="open").order_by(index=pkey).run(conn, time_format="raw", binary_format='raw')
+        table = table.get_all(*keys).order_by(pkey)
```

Comment:
@chrisvariety I'm not sure I understand. The query uses `between` when there is a `progress[0]` in order to resume after reconnection. Then we optionally do a `get_all` to specify a subset of documents, and then an `order_by` on the primary key (using an index only when selecting all documents). Processing the documents in primary key order ensures that the export progress is predictable.

As far as I can see, the only issue is that `order_by` cannot use the index when we are selecting a subset using `get_all`, but this is inescapable.